tests/database/test_eth1_chaindb.py

```diff
@@ -84,13 +88,153 @@ def chain(chain_without_block_validation):
+
+    gap_chain = api.copy(common_chain)
+    assert gap_chain.get_canonical_head() == common_chain.get_canonical_head()
+
+    gap_chain.chaindb.persist_checkpoint_header(main_header_6, main_header_6_score)
+    # We created a gap in the chain of headers
+    assert gap_chain.chaindb.get_header_chain_gaps() == (((4, 5),), 7)
+    # ...but not in the chain of blocks (yet!)
+    assert gap_chain.chaindb.get_chain_gaps() == ((), 4)
+    block_7 = main_chain.get_canonical_block_by_number(7)
+    block_7_receipts = block_7.get_receipts(main_chain.chaindb)
+    # Persist block 7 on top of the checkpoint
+    gap_chain.chaindb.persist_unexecuted_block(block_7, block_7_receipts)
+    assert gap_chain.chaindb.get_header_chain_gaps() == (((4, 5),), 8)
+    # Now we have a gap in the chain of blocks, too
+    assert gap_chain.chaindb.get_chain_gaps() == (((4, 6),), 8)
+
+    # Overwriting header 3 doesn't cause us to re-open a block gap
+    gap_chain.chaindb.persist_header_chain([
+        main_chain.chaindb.get_canonical_block_header_by_number(3)
+    ])
+    assert gap_chain.chaindb.get_chain_gaps() == (((4, 6),), 8)
+
+    # Now get the uncle block
+    uncle_block = uncle_chain.get_canonical_block_by_number(4)
+    uncle_block_receipts = uncle_block.get_receipts(uncle_chain.chaindb)
+
+    # Put the uncle block in the gap
+    gap_chain.chaindb.persist_unexecuted_block(uncle_block, uncle_block_receipts)
+    assert gap_chain.chaindb.get_header_chain_gaps() == (((5, 5),), 8)
+    assert gap_chain.chaindb.get_chain_gaps() == (((5, 6),), 8)
+
+    # Trying to save another uncle errors as its header isn't the parent of the checkpoint
+    second_uncle = uncle_chain.get_canonical_block_by_number(5)
+    second_uncle_receipts = second_uncle.get_receipts(uncle_chain.chaindb)
+    with pytest.raises(CheckpointsMustBeCanonical):
+        gap_chain.chaindb.persist_unexecuted_block(second_uncle, second_uncle_receipts)
+
+    # Now close the gap in the header chain with the actual correct headers
+    actual_headers = [
+        main_chain.chaindb.get_canonical_block_header_by_number(block_number)
+        for block_number in range(4, 7)
+    ]
+    gap_chain.chaindb.persist_header_chain(actual_headers)
+    # No more gaps in the header chain
+    assert gap_chain.chaindb.get_header_chain_gaps() == ((), 8)
+    # We detected the de-canonicalized uncle and re-opened the block gap
+    assert gap_chain.chaindb.get_chain_gaps() == (((4, 6),), 8)
+
+    # The uncle block was implicitly de-canonicalized with its header,
+    # hence we can not fetch it any longer.
+    if can_fetch_block:
+        block_4 = gap_chain.get_canonical_block_by_number(4)
+        assert block_4 == main_chain.get_canonical_block_by_number(4)
+    else:
+        with pytest.raises(BlockNotFound):
+            gap_chain.get_canonical_block_by_number(4)
+        # Add the missing block and assert the gap shrinks
+        assert gap_chain.chaindb.get_chain_gaps() == (((4, 6),), 8)
```

Comment:
This line is redundant because nothing has changed since the last time we asserted the same thing. But if you like it being redundant for the comparison to line 235, I don't really object. Just making sure it's not accidental code cruft.

Reply:
Yeah, I put it in on purpose just to make the reader aware again of the current situation in terms of gaps.